### PR TITLE
init: multiple cleanups

### PR DESCRIPTION
**Pull request by gmarull**
Similar to the initial work done for `device.h` in https://github.com/zephyrproject-rtos/zephyr/pull/50827, now it's the turn for `init.h`.

Summary of the changes:

- Improve Doxygen
- Add some auxiliary macros to improve code readability
- z_sys_init_run_level is no longer sitting in a public header, same as associated internal defines

Note: please review on a per-commit basis.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

### CODE REVIEW

File: include/zephyr/init.h
Comment (nashif) [inline include/zephyr/init.h:68]: pleasee 9247e8bc44145c88654d54d8b380bc9514e2c514, those underscores were added to deal with some coding guideline violation, need to check if this still applies.

Comment (gmarull) [reply include/zephyr/init.h:68]: I can't see any immediate violation of 5.7 rule in `init.h` any expert on the topic that can review? In fact, I'm not sure what was fixed in 9247e8b regarding init.h (commit message doesn't mention why `_` was added in init.h or device.h...).

File: include/zephyr/init.h
Comment (carlocaione) [inline include/zephyr/init.h:100]: nit: are `\` tab-aligned?

**Review by carlocaione — DISMISSED** [review]
Al long as CI is happy, I'm happy too.

File: include/zephyr/init.h
Comment (gmarull) [reply include/zephyr/init.h:100]: > nit: are `\` tab-aligned?

not if using clang-format

**Review by henrikbrixandersen — DISMISSED** [review]
Nice clean-up. Thank you.

**Review by carlescufi — DISMISSED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by stephanosio — DISMISSED** [review]
There is a conflict now:
https://github.com/zephyrproject-rtos/zephyr/commit/e8395351e66b2c83ff4227df1f4c9f1f31d28f0c

File: include/zephyr/init.h
Comment (stephanosio) [inline include/zephyr/init.h:27]: what about `ARCH`?

File: include/zephyr/init.h
Comment (stephanosio) [inline include/zephyr/init.h:114]: `ARCH`?

**Review by stephanosio — CHANGES_REQUESTED** [review]
what do we do about:

https://github.com/zephyrproject-rtos/zephyr/blob/3db1c59f444f3a090d5d96910001a14132137708/soc/xtensa/intel_adsp/common/boot.c#L155

It looks like `z_sys_init_run_level` should be part of `init.h` after all.

**Comment by gmarull** [comment]
> what do we do about:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/3db1c59f444f3a090d5d96910001a14132137708/soc/xtensa/intel_adsp/common/boot.c#L155
> 
> It looks like `z_sys_init_run_level` should be part of `init.h` after all.

How did we end up merging stuff like that without proper review of the device model maintainers?

File: include/zephyr/init.h
Comment (gmarull) [reply include/zephyr/init.h:27]: I will add it, apparently it was added without any documentation (sigh).

Comment (nashif) [reply include/zephyr/init.h]: added https://github.com/zephyrproject-rtos/zephyr/pull/51176

**Review by carlescufi — DISMISSED** [review]

**Comment by gmarull** [comment]
> > How did we end up merging stuff like that without proper review of the device model maintainers?
> 
> what is the problem exactly with "stuff like that"?

How can we modify core infrastructure like init levels without reviews from the maintainers? To start with, `ARCH` level was added with no documentation at all. How am I supposed to use it? When is it supposed to run? What are its guarantees? Is this really an Intel-specific thing that quietly got in?

File: kernel/init.c
Comment (stephanosio) [inline kernel/init.c:499]: I do not think the `ARCH` level is supposed to do that based on the PR that introduced it; but, I agree it is quite vague in its current form.

Comment (gmarull) [reply kernel/init.c:499]: The problem is that PR introducing such level was merged without proper reviewing, no docs, etc. So, who knows what "ARCH" init level means now.

**Review by andyross — DISMISSED** [review]
One nitpick, but not worth a -1.  Everything else looks great to me.

**Comment by andyross** [comment]
Maybe remove the init level code from this PR?  We can come back later with a cleaner design that scales better.  Having an ARCH layer definitely makes some sense, lots of devices have code that needs to run before z_cstart() and there's probably some value in providing SYS_INIT() to them for that purpose.

**Review by stephanosio — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

File: subsys/shell/modules/device_service.c
Comment (tbursztyka) [reply subsys/shell/modules/device_service.c:35]: I don't see why enum init_level cannot be public, here for instance it misses the EARLY and makes this confusing.

**Review by tbursztyka — APPROVED** [review]
lgtm, just one comment on init level

File: subsys/shell/modules/device_service.c
Comment (gmarull) [reply subsys/shell/modules/device_service.c:35]: Precisely because it misses EARLY the code actually works now (`main` is currently broken). These are indices of an array internal to `z_sys_init_run_level`, not public anymore after these changes. The device service shell creates its own array, and as explained in the commit, it's up to the module on how to handle access to such array. They should only be public if used as identifiers, not tied to any implementation indexing technique.

**Pull request merged**
